File: 563-find-tilt/solution.go
package main
import (
	"gopractice/datastructure/Tree"
)

type TreeNode = Tree.TreeNode

func findTilt(root *TreeNode) int {
	result := 0
	dfs(root, &result)
	return result
}
func dfs(node *TreeNode, result *int) int {
	if node == nil{
		return  0
	}	
	// 因為 result 在這裡已經是指標 (*int) 了，直接傳下去即可，不用再加 &
	leftSum := dfs(node.Left, result)
	rightSum := dfs(node.Right, result)
	slope := abs(leftSum - rightSum)
	*result += slope
	return leftSum + rightSum + node.Val.(int)
}

func abs(x int) int {
    if x < 0 {
        return -x
    }
    return x
}

/*
*result += slope：若要修改指標指向的變數數值，必須在變數前加上 *（解引用，Dereference）。  

result 儲存的是記憶體地址。
*result 代表該地址裡實際存放的整數。因此 *result += slope 就能真正改到 findTilt 裡的 result。
*/

File: 563-find-tilt/solution.js
const { range } = require('balanced-match');
const TreeNode = require('../datastructure/Tree/TreeNode');
const TreeDebugger = require('../datastructure/Tree/TreeDebugger')
/**
 * @param {TreeNode} root
 * @return {number}
 */
var findTilt = function(root) {
    let result = 0
    function dfs(node){
        if(!node){
            return 0
        }
        let leftSum = dfs(node.left)
        let rightSum = dfs(node.right)
        slope = Math.abs(leftSum - rightSum)
        result += slope
        return leftSum + rightSum + node.val
    }
    dfs(root)
    return result
};
if(require.main === module){
    const levelOrder = [1,2,3]
    const root = TreeDebugger.buildLevelOrderTree(levelOrder)
    console.log("Ans = ", findTilt(root))
}